instrumentation/jdbc/library/src/main/java/io/opentelemetry/instrumentation/jdbc/internal/OpenTelemetryPreparedStatement.java

```diff
@@ -0,0 +1,397 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// Includes work from:
+/*
+ * Copyright 2017-2021 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.opentelemetry.instrumentation.jdbc.internal;
+
+import static io.opentelemetry.instrumentation.jdbc.internal.JdbcSingletons.instrumenter;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.javaagent.instrumentation.api.CallDepthThreadLocalMap;
+import java.io.InputStream;
+import java.io.Reader;
+import java.math.BigDecimal;
+import java.net.URL;
+import java.sql.Array;
+import java.sql.Blob;
+import java.sql.Clob;
+import java.sql.Date;
+import java.sql.NClob;
+import java.sql.ParameterMetaData;
+import java.sql.PreparedStatement;
+import java.sql.Ref;
+import java.sql.ResultSet;
+import java.sql.ResultSetMetaData;
+import java.sql.RowId;
+import java.sql.SQLException;
+import java.sql.SQLXML;
+import java.sql.Statement;
+import java.sql.Time;
+import java.sql.Timestamp;
+import java.util.Calendar;
+
+class OpenTelemetryPreparedStatement<S extends PreparedStatement> extends OpenTelemetryStatement<S>
+    implements PreparedStatement {
+
+  public OpenTelemetryPreparedStatement(S preparedStatement, String query) {
+    super(preparedStatement, query);
+  }
+
+  private static <T, E extends Exception> T wrapCall(
+      PreparedStatement preparedStatement, CheckedCallable<T, E> callable) throws E {
+    // Connection#getMetaData() may execute a Statement or PreparedStatement to retrieve DB info
+    // this happens before the DB CLIENT span is started (and put in the current context), so this
+    // instrumentation runs again and the shouldStartSpan() check always returns true - and so on
+    // until we get a StackOverflowError
+    // using CallDepth prevents this, because this check happens before Connection#getMetadata()
+    // is called - the first recursive Statement call is just skipped and we do not create a span
+    // for it
+    if (CallDepthThreadLocalMap.getCallDepth(Statement.class).getAndIncrement() > 0) {
+      return callable.call();
+    }
```

Comment:
I wonder if this is needed for the library instrumentation: this bug only manifested itself when instrumented applications were creating JDBC connections directly (`new SomeDbConnection(...).getMetadata()`), without calling a driver. Since this instrumentation wraps over everything there may be no need for this code (and that'll help getting rid of the `javaagent-api` dependency which shouldn't be present in a library instrumentation)

Reply:
Removed it.